asserts/headers.go

```diff
@@ -0,0 +1,129 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"unicode/utf8"
+)
+
+func parseHeaders(head []byte) (map[string]interface{}, error) {
+	if !utf8.Valid(head) {
+		return nil, fmt.Errorf("header is not utf8")
+	}
+	headers := make(map[string]interface{})
+	lines := strings.Split(string(head), "\n")
+	for i := 0; i < len(lines); {
+		entry := lines[i]
+		nameValueSplit := strings.Index(entry, ":")
+		if nameValueSplit == -1 {
+			return nil, fmt.Errorf("header entry missing ':' separator: %q", entry)
+		}
+		name := entry[:nameValueSplit]
+		if !headerNameSanity.MatchString(name) {
+			return nil, fmt.Errorf("invalid header name: %q", name)
+		}
+
+		consumed := nameValueSplit + 1
+		var value interface{}
+		var err error
+		value, i, err = parseEntry(consumed, i, lines, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		headers[name] = value
+	}
+	return headers, nil
+}
+
+func nestingPrefix(nesting int, prefix string) string {
+	return strings.Repeat("    ", nesting) + prefix
+}
+
+func parseEntry(consumedByIntro int, i int, lines []string, nesting int) (value interface{}, after int, err error) {
+	entry := lines[i]
+	i++
+	if consumedByIntro == len(entry) {
+		// multiline values
+		if i < len(lines) && strings.HasPrefix(lines[i], nestingPrefix(nesting, "  -")) {
+			// list
+			return parseList(i, lines, nesting)
+		}
+
+		return parseMultilineText(i, lines, nesting)
+	}
+
+	// simple one-line value
+	if entry[consumedByIntro] != ' ' {
+		return nil, -1, fmt.Errorf("header entry should have a space or newline (multiline) before value: %q", entry)
```

Comment:
might be nice to add the number of the failing line to this error, but it only really matters once we have nested maps so is not too important right now.